File: src/main/java/com/sfh/pokeRogueBot/service/Brain.java
package com.sfh.pokeRogueBot.service;

import com.sfh.pokeRogueBot.model.decisions.*;
import com.sfh.pokeRogueBot.model.dto.SaveSlotDto;
import com.sfh.pokeRogueBot.model.dto.WaveDto;
import com.sfh.pokeRogueBot.model.enums.CommandPhaseDecision;
import com.sfh.pokeRogueBot.model.enums.RunStatus;
import com.sfh.pokeRogueBot.model.enums.UiMode;
import com.sfh.pokeRogueBot.model.exception.StopRunException;
import com.sfh.pokeRogueBot.model.modifier.ChooseModifierItem;
import com.sfh.pokeRogueBot.model.modifier.ModifierShop;
import com.sfh.pokeRogueBot.model.modifier.MoveToModifierResult;
import com.sfh.pokeRogueBot.model.poke.Pokemon;
import com.sfh.pokeRogueBot.model.run.RunProperty;
import com.sfh.pokeRogueBot.neurons.*;
import com.sfh.pokeRogueBot.phase.NoUiPhase;
import com.sfh.pokeRogueBot.phase.Phase;
import com.sfh.pokeRogueBot.phase.ScreenshotClient;
import com.sfh.pokeRogueBot.phase.UiPhase;
import com.sfh.pokeRogueBot.service.javascript.JsService;
import com.sfh.pokeRogueBot.service.javascript.JsUiService;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import java.util.List;

@Slf4j
@Service
public class Brain {

    private final JsService jsService;
    private final JsUiService jsUiService;
    private final ShortTermMemory shortTermMemory;
    private final LongTermMemory longTermMemory;
    private final ScreenshotClient screenshotClient;

    private final SwitchPokemonNeuron switchPokemonNeuron;
    private final ChooseModifierNeuron chooseModifierNeuron;
    private final CombatNeuron combatNeuron;
    private final CapturePokemonNeuron capturePokemonNeuron;
    private final LearnMoveNeuron learnMoveNeuron;

    private RunProperty runProperty = null;
    private boolean waveIndexReset = false;
    private WaveDto waveDto;
    private ChooseModifierDecision chooseModifierDecision;
    @Getter
    private SaveSlotDto[] saveSlots;

    public Brain(
            JsService jsService,
            JsUiService jsUiService,
            ShortTermMemory shortTermMemory,
            LongTermMemory longTermMemory,
            ScreenshotClient screenshotClient,
            SwitchPokemonNeuron switchPokemonNeuron,
            ChooseModifierNeuron chooseModifierNeuron,
            CombatNeuron combatNeuron,
            CapturePokemonNeuron capturePokemonNeuron,
            LearnMoveNeuron learnMoveNeuron
    ) {
        this.jsService = jsService;
        this.jsUiService = jsUiService;
        this.shortTermMemory = shortTermMemory;
        this.longTermMemory = longTermMemory;
        this.screenshotClient = screenshotClient;
        this.switchPokemonNeuron = switchPokemonNeuron;
        this.chooseModifierNeuron = chooseModifierNeuron;
        this.combatNeuron = combatNeuron;
        this.capturePokemonNeuron = capturePokemonNeuron;
        this.learnMoveNeuron = learnMoveNeuron;
    }

    public SwitchDecision getPokemonSwitchDecision(boolean ignoreFirstPokemon) {
        waveDto = jsService.getWaveDto(); //always update current state
        return switchPokemonNeuron.getBestSwitchDecision(waveDto, ignoreFirstPokemon);
    }

    public MoveToModifierResult getModifierToPick() {
        if (null == chooseModifierDecision) { //get new decision
            this.waveDto = jsService.getWaveDto(); //always refresh money and pokemons before choosing the modifiers
            ModifierShop shop = jsUiService.getModifierShop();
            List<ChooseModifierItem> allItems = shop.getAllItems();
            longTermMemory.memorizeItems(allItems);
            this.chooseModifierDecision = chooseModifierNeuron.getModifierToPick(waveDto.getWavePokemon().getPlayerParty(), waveDto, shop);
        }

        if (!chooseModifierDecision.getItemsToBuy().isEmpty()) { //buy items first
            MoveToModifierResult result = chooseModifierDecision.getItemsToBuy().get(0);
            chooseModifierDecision.getItemsToBuy().remove(0);
            return result;
        }

        MoveToModifierResult freeItem = chooseModifierDecision.getFreeItemToPick();
        chooseModifierDecision = null; //reset to null, so that in the next wave a new decision is taken
        return freeItem;  //if no items have to be bought, pick a free item
    }

    public CommandPhaseDecision getCommandDecision() {
        waveDto = jsService.getWaveDto(); //always update current state

        return CommandPhaseDecision.ATTACK;
    }

    public AttackDecision getAttackDecision() {

        waveDto = jsService.getWaveDto(); //always update current state

        if (waveDto.isDoubleFight()) {

            Pokemon[] playerParty = waveDto.getWavePokemon().getPlayerParty();
            Pokemon[] enemyParty = waveDto.getWavePokemon().getEnemyParty();

            int playerPartySize = 0;
            for (Pokemon pokemon : playerParty) {
                if (pokemon.getHp() > 0) {
                    playerPartySize++;
                }
            }
            int enemyPartySize = 0;
            for (Pokemon pokemon : enemyParty) {
                if (pokemon.getHp() > 0) {
                    enemyPartySize++;
                }
            }

            Pokemon playerPokemon1 = playerParty[0].isAlive() ? playerParty[0] : null;
            Pokemon playerPokemon2 = playerPartySize > 0 && playerParty[1].isAlive() ? playerParty[1] : null;

            Pokemon enemyPokemon1 = enemyParty[0].isAlive() ? enemyParty[0] : null;
            Pokemon enemyPokemon2 = enemyPartySize > 0 && enemyParty[1].isAlive() ? enemyParty[1] : null;

            AttackDecisionForDoubleFight forDoubleFight = combatNeuron.getAttackDecisionForDoubleFight(
                    playerPokemon1,
                    playerPokemon2,
                    enemyPokemon1,
                    enemyPokemon2
            );
            return forDoubleFight;
        } else {
            //single fight
            Pokemon wildPokemon = waveDto.getWavePokemon().getEnemyParty()[0];
            return combatNeuron.getAttackDecisionForSingleFight(
                    waveDto.getWavePokemon().getPlayerParty()[0],
                    wildPokemon,
                    capturePokemonNeuron.shouldCapturePokemon(waveDto, wildPokemon)
            );
        }
    }

    public int selectStrongestPokeball() {
        return capturePokemonNeuron.selectStrongestPokeball(waveDto);
    }

    public void informWaveEnded(int newWaveIndex) {
        this.waveDto = jsService.getWaveDto();
        this.chooseModifierDecision = null;
        runProperty.setWaveIndex(newWaveIndex);
        runProperty.updateTeamSnapshot(waveDto.getWavePokemon().getPlayerParty());
        runProperty.setMoney(waveDto.getMoney());
        log.debug("new wave: Waveindex: " + waveDto.getWaveIndex() + ", is trainer fight: " + waveDto.isTrainerFight());

        if (null != waveDto && waveDto.isWildPokemonFight()) {
            for (Pokemon wildPokemon : waveDto.getWavePokemon().getEnemyParty()) {
                if (wildPokemon.isShiny()) {
                    String message = "Shiny pokemon detected: " + wildPokemon.getName();
                    log.info(message);
                    screenshotClient.persistScreenshot("shiny_pokemon_detected");
                    throw new StopRunException(message);
                }
                if (wildPokemon.getSpecies().getMythical()) {
                    String message = "Mythical pokemon detected: " + wildPokemon.getName();
                    log.info(message);
                    screenshotClient.persistScreenshot("mythical_pokemon_detected");
                    throw new StopRunException(message);
                }
                if (wildPokemon.getSpecies().getLegendary()) {
                    String message = "Legendary pokemon detected: " + wildPokemon.getName();
                    log.info(message);
                    screenshotClient.persistScreenshot("legendary_pokemon_detected");
                    throw new StopRunException(message);
                }
                if (wildPokemon.getSpecies().getSubLegendary()) {
                    String message = "Sub Legendary pokemon detected: " + wildPokemon.getName();
                    log.info(message);
                    screenshotClient.persistScreenshot("sub_legendary_pokemon_detected");
                    throw new StopRunException(message);
                }
            }
        }
    }

    public void memorize(String phase) {
        shortTermMemory.memorizePhase(phase);
    }

    public void clearShortTermMemory() {
        shortTermMemory.clearLastPhaseMemory();
    }

    public void rememberLongTermMemories() {
        log.debug("Remember long term memories");
        longTermMemory.rememberItems();
        longTermMemory.rememberUiValidatedPhases();
    }

    /**
     * If the save slots are not loaded, open the save slots menu to get the save slots data
     * If the save slots are loaded, check if there is a save slot without an error
     *
     * @return if the load game menu should be opened
     */
    public boolean shouldLoadGame() {
        if (null == saveSlots) {
            return true;
        }

        for (SaveSlotDto saveSlot : saveSlots) {
            if (saveSlot.isDataPresent() && !saveSlot.isErrorOccurred()) {
                return true;
            }
        }

        return false;
    }

    /**
     * When called, the save slot menu should be opened so the data is accessible with JS
     *
     * @return which save slot index should be loaded or -1 if no save slot should be loaded
     */
    public int getSaveSlotIndexToLoad() {
        if (null == saveSlots) {
            this.saveSlots = jsUiService.getSaveSlots();
        }

        for (SaveSlotDto saveSlot : saveSlots) {
            if (saveSlot.isDataPresent() && !saveSlot.isErrorOccurred()) {
                return saveSlot.getSlotId();
            }
        }

        return -1;
    }

    /**
     * When called, the save slot data are already loaded.
     * Returns which save slot index should be saved to or -1 if no save slot is free.
     *
     * @return a value >= 0 if the save slot is empty and has no error
     */
    public int getSaveSlotIndexToSave() {
        if (null == saveSlots) {
            throw new IllegalStateException("Save slots are not loaded, cannot determine save slot index to save");
        }

        for (SaveSlotDto saveSlot : saveSlots) {
            if (!saveSlot.isDataPresent() && !saveSlot.isErrorOccurred()) {
                return saveSlot.getSlotId();
            }
        }

        return -1;
    }

    public RunProperty getRunProperty() {
        if (runProperty == null) {
            log.debug("runProperty is null, creating new one");
            runProperty = new RunProperty(1);
            waveIndexReset = true;
            return runProperty;
        }

        if (runProperty.getStatus() == RunStatus.OK) {
            log.debug("runProperty is OK, returning runProperty");
            return runProperty;
        }

        if (null == saveSlots) {
            throw new IllegalStateException("Save slots are not loaded, cannot determine run property");
        }

        switch (runProperty.getStatus()) {
            case ERROR, RELOAD_APP:
                if (runProperty.getSaveSlotIndex() != -1) {
                    log.debug("Error occurred, setting error to save slot: " + runProperty.getSaveSlotIndex());
                    saveSlots[runProperty.getSaveSlotIndex()].setErrorOccurred(true);
                } else {
                    log.debug("Save slot index is -1, so error occurred before starting a run.");
                }
                runProperty = new RunProperty(runProperty.getRunNumber() + 1);
                waveIndexReset = true;
                return runProperty;
            case LOST:
                log.debug("Lost battle, setting data present to false for save slot: " + runProperty.getSaveSlotIndex());
                saveSlots[runProperty.getSaveSlotIndex()].setErrorOccurred(false);
                saveSlots[runProperty.getSaveSlotIndex()].setDataPresent(false);
                runProperty = new RunProperty(runProperty.getRunNumber() + 1);
                waveIndexReset = true;
                return runProperty;
            default:
                throw new IllegalStateException("RunProperty has unknown status: " + runProperty.getStatus());
        }
    }

    public boolean tryToCatchPokemon() {
        waveDto = jsService.getWaveDto(); //always update current state
        return capturePokemonNeuron.shouldCapturePokemon(waveDto, waveDto.getWavePokemon().getEnemyParty()[0]);
    }

    public SwitchDecision getBestSwitchDecision() {
        SwitchDecision switchDecision = switchPokemonNeuron.getBestSwitchDecision(waveDto, false);
        if (switchDecision == null) {
            throw new IllegalStateException("No switch decision found");
        }
        log.debug("Switching to pokemon: " + switchDecision.getPokeName() + " on index: " + switchDecision.getIndex());
        return switchDecision;
    }

    public boolean shouldSwitchPokemon() {
        waveDto = jsService.getWaveDto(); //always update current state
        return switchPokemonNeuron.shouldSwitchPokemon(waveDto);
    }

    public LearnMoveDecision getLearnMoveDecision(Pokemon pokemon) {
        return learnMoveNeuron.getLearnMoveDecision(pokemon);
    }

    public boolean shouldResetwaveIndex() {
        if (waveIndexReset) {
            waveIndexReset = false;
            return true;
        }
        return false;
    }

    @Deprecated()
    public boolean phaseUiIsValidated(@NotNull Phase phase, @NotNull UiMode uiMode) {
        boolean isValidated = longTermMemory.isUiValidated(phase);
        if (isValidated) {
            return true;
        }
        if (phase instanceof NoUiPhase noUiPhase) {
            longTermMemory.memorizePhase(noUiPhase.getPhaseName());
            return true;
        }
        if (phase instanceof UiPhase uiPhase) {
            longTermMemory.memorizePhase(uiPhase.getPhaseName());
            return true;
        }
        return false; //missing Interface on PhaseClass and developer has to check the phase in the browser
    }
}

File: src/main/kotlin/com/sfh/pokeRogueBot/service/Brain.kt
package com.sfh.pokeRogueBot.service

import com.sfh.pokeRogueBot.model.decisions.AttackDecision
import com.sfh.pokeRogueBot.model.decisions.LearnMoveDecision
import com.sfh.pokeRogueBot.model.decisions.SwitchDecision
import com.sfh.pokeRogueBot.model.dto.SaveSlotDto
import com.sfh.pokeRogueBot.model.dto.WaveDto
import com.sfh.pokeRogueBot.model.enums.CommandPhaseDecision
import com.sfh.pokeRogueBot.model.enums.RunStatus
import com.sfh.pokeRogueBot.model.enums.UiMode
import com.sfh.pokeRogueBot.model.modifier.MoveToModifierResult
import com.sfh.pokeRogueBot.model.poke.Pokemon
import com.sfh.pokeRogueBot.model.rl.*
import com.sfh.pokeRogueBot.model.run.RunProperty
import com.sfh.pokeRogueBot.neurons.*
import com.sfh.pokeRogueBot.phase.NoUiPhase
import com.sfh.pokeRogueBot.phase.Phase
import com.sfh.pokeRogueBot.phase.UiPhase
import com.sfh.pokeRogueBot.service.javascript.JsService
import com.sfh.pokeRogueBot.service.javascript.JsUiService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class Brain(
    private val jsService: JsService,
    private val jsUiService: JsUiService,
    private val shortTermMemory: ShortTermMemory,
    private val longTermMemory: LongTermMemory,
    private val switchPokemonNeuron: SwitchPokemonNeuron,
    private val combatNeuron: CombatNeuron,
    private val capturePokemonNeuron: CapturePokemonNeuron,
    private val learnMoveNeuron: LearnMoveNeuron,
    private val modifierRLNeuron: ModifierRLNeuron,
    private val rewardCalculator: ModifierRewardCalculator
) {

    companion object {
        private val log = LoggerFactory.getLogger(Brain::class.java)
    }

    private var runProperty: RunProperty? = null
    private var waveIndexReset = false
    private lateinit var waveDto: WaveDto
    private var saveSlots: Array<SaveSlotDto>? = null

    // Episode-based RL components
    private val episodeManager = ModifierEpisodeManager()
    private val decisionLogger = ModifierDecisionLogger()
    private var currentModifierState: SmallModifierSelectState? = null

    fun getPokemonSwitchDecision(ignoreFirstPokemon: Boolean): SwitchDecision {
        waveDto = jsService.getWaveDto()
        return switchPokemonNeuron.getBestSwitchDecision(waveDto, ignoreFirstPokemon)
    }

    /**
     * 1. Get State
     * 2. Get Available Actions
     * 3. RL Agent selects action
     * 4. Convert action to game result
     * 5. Log experience for training
     */
    fun getModifierToPick(): MoveToModifierResult? {
        this.waveDto = jsService.getWaveDto()
        val shop = jsUiService.getModifierShop()

        // Step 1: Create state for RL logging
        currentModifierState = SmallModifierSelectState.create(
            pokemons = waveDto.wavePokemon.playerParty,
            shopItems = shop.shopItems,
            freeItems = shop.freeItems,
            currentMoney = waveDto.money
        )
        longTermMemory.memorizeItems(shop.allItems)

        // Step 2: Get available actions for RL agent
        val availableActions = modifierRLNeuron.getAvailableActions(
            shop = shop,
            currentMoney = waveDto.money,
            playerParty = waveDto.wavePokemon.playerParty
        )

        // Step 3: RL agent selects action
        val selectedAction = modifierRLNeuron.selectAction(
            state = currentModifierState!!,
            availableActions = availableActions
        )

        // Step 4: Convert RL action to game-executable result
        val result = modifierRLNeuron.convertActionToResult(
            action = selectedAction,
            shop = shop,
            team = waveDto.wavePokemon.playerParty
        )

        // Step 5: Add step to current RL episode
        addStepToCurrentEpisode(selectedAction, currentModifierState!!, waveDto.waveIndex)

        log.info(
            "RL modifier decision: action={}, result={}, availableActions={}",
            selectedAction, result?.toString() ?: "SKIP", availableActions
        )

        return result
    }

    fun getCommandDecision(): CommandPhaseDecision {
        waveDto = jsService.getWaveDto()
        return CommandPhaseDecision.ATTACK
    }

    fun getAttackDecision(): AttackDecision? {
        waveDto = jsService.getWaveDto()

        return if (waveDto.isDoubleFight) {
            val playerParty = waveDto.wavePokemon.playerParty
            val enemyParty = waveDto.wavePokemon.enemyParty

            val playerPartySize = playerParty.count { it.hp > 0 }
            val enemyPartySize = enemyParty.count { it.hp > 0 }

            val playerPokemon1 = if (playerParty[0].isAlive()) playerParty[0] else null
            val playerPokemon2 = if (playerPartySize > 0 && playerParty[1].isAlive()) playerParty[1] else null

            val enemyPokemon1 = if (enemyParty[0].isAlive()) enemyParty[0] else null
            val enemyPokemon2 = if (enemyPartySize > 0 && enemyParty[1].isAlive()) enemyParty[1] else null

            combatNeuron.getAttackDecisionForDoubleFight(
                playerPokemon1,
                playerPokemon2,
                enemyPokemon1,
                enemyPokemon2
            )
        } else {
            val wildPokemon = waveDto.wavePokemon.enemyParty.firstOrNull()
                ?: throw IllegalStateException("No enemy pokemon found in enemyParty")

            combatNeuron.getAttackDecisionForSingleFight(
                waveDto.wavePokemon.playerParty[0],
                wildPokemon,
                capturePokemonNeuron.shouldCapturePokemon(waveDto, wildPokemon)
            )
        }
    }

    fun selectStrongestPokeball(): Int {
        return capturePokemonNeuron.selectStrongestPokeball(waveDto)
    }

    fun informWaveEnded(newWaveIndex: Int) {
        this.waveDto = jsService.getWaveDto()

        // Episode continues - no action needed here
        // Episodes are completed only when runs end (team wipe/victory)

        // chooseModifierDecision cleared (removed old field reference)
        runProperty!!.waveIndex = newWaveIndex
        runProperty!!.updateTeamSnapshot(waveDto.wavePokemon.playerParty)
        runProperty!!.money = waveDto.money
        log.debug("new wave: Waveindex: ${waveDto.waveIndex}, is trainer fight: ${waveDto.isTrainerFight()}")
    }

    fun memorize(phase: String) {
        shortTermMemory.memorizePhase(phase)
    }

    fun clearShortTermMemory() {
        shortTermMemory.clearLastPhaseMemory()
    }

    fun rememberLongTermMemories() {
        log.debug("Remember long term memories")
        longTermMemory.rememberItems()
        longTermMemory.rememberUiValidatedPhases()
    }

    /**
     * If the save slots are not loaded, open the save slots menu to get the save slots data
     * If the save slots are loaded, check if there is a save slot without an error
     *
     * @return if the load game menu should be opened
     */
    fun shouldLoadGame(): Boolean {
        if (saveSlots == null) {
            return true
        }

        for (saveSlot in saveSlots) {
            if (saveSlot.isDataPresent && !saveSlot.isErrorOccurred) {
                return true
            }
        }

        return false
    }

    /**
     * When called, the save slot menu should be opened so the data is accessible with JS
     *
     * @return which save slot index should be loaded or -1 if no save slot should be loaded
     */
    fun getSaveSlotIndexToLoad(): Int {
        if (saveSlots == null) {
            this.saveSlots = jsUiService.getSaveSlots()
        }

        for (saveSlot in saveSlots!!) {
            if (saveSlot.isDataPresent && !saveSlot.isErrorOccurred) {
                return saveSlot.slotId
            }
        }

        return -1
    }

    /**
     * When called, the save slot data are already loaded.
     * Returns which save slot index should be saved to or -1 if no save slot is free.
     *
     * @return a value >= 0 if the save slot is empty and has no error
     */
    fun getSaveSlotIndexToSave(): Int {
        if (saveSlots == null) {
            throw IllegalStateException("Save slots are not loaded, cannot determine save slot index to save")
        }

        for (saveSlot in saveSlots!!) {
            if (!saveSlot.isDataPresent && !saveSlot.isErrorOccurred) {
                return saveSlot.slotId
            }
        }

        return -1
    }

    fun getRunProperty(): RunProperty {
        if (runProperty == null) {
            log.debug("runProperty is null, creating new one")
            runProperty = RunProperty(1)
            waveIndexReset = true

            return runProperty!!
        }

        if (runProperty!!.status == RunStatus.OK) {
            log.debug("runProperty is OK, returning runProperty")
            return runProperty!!
        }

        if (saveSlots == null) {
            throw IllegalStateException("Save slots are not loaded, cannot determine run property")
        }

        return when (runProperty!!.status) {
            RunStatus.ERROR, RunStatus.RELOAD_APP -> {
                if (runProperty!!.saveSlotIndex != -1) {
                    log.debug("Error occurred, setting error to save slot: ${runProperty!!.saveSlotIndex}")
                    saveSlots!![runProperty!!.saveSlotIndex].isErrorOccurred = true
                    // Complete current RL episode with error outcome
                    completeCurrentEpisode(RunTerminalOutcome.ERROR_OCCURRED, waveDto.waveIndex)
                } else {
                    log.debug("Save slot index is -1, so error occurred before starting a run.")
                }
                runProperty = RunProperty(runProperty!!.runNumber + 1)
                waveIndexReset = true
                startNewRLEpisode(false) // Start new episode for the restart (fresh run)
                runProperty!!
            }

            RunStatus.LOST -> {
                log.debug("Lost battle, setting data present to false for save slot: ${runProperty!!.saveSlotIndex}")

                // Complete current RL episode with team wipe outcome
                completeCurrentEpisode(RunTerminalOutcome.TEAM_WIPE, waveDto.waveIndex)

                saveSlots!![runProperty!!.saveSlotIndex].isErrorOccurred = false
                saveSlots!![runProperty!!.saveSlotIndex].isDataPresent = false
                runProperty = RunProperty(runProperty!!.runNumber + 1)
                waveIndexReset = true
                runProperty!!
            }

            else -> throw IllegalStateException("RunProperty has unknown status: ${runProperty!!.status}")
        }
    }

    fun tryToCatchPokemon(): Boolean {
        waveDto = jsService.getWaveDto()
        return capturePokemonNeuron.shouldCapturePokemon(waveDto, waveDto.wavePokemon?.enemyParty?.get(0))
    }

    fun getBestSwitchDecision(): SwitchDecision {
        val switchDecision = switchPokemonNeuron.getBestSwitchDecision(waveDto, false)
            ?: throw IllegalStateException("No switch decision found")
        log.debug("Switching to pokemon: ${switchDecision.pokeName} on index: ${switchDecision.index}")
        return switchDecision
    }

    fun shouldSwitchPokemon(): Boolean {
        waveDto = jsService.getWaveDto()
        return switchPokemonNeuron.shouldSwitchPokemon(waveDto)
    }

    fun getLearnMoveDecision(pokemon: Pokemon): LearnMoveDecision {
        return learnMoveNeuron.getLearnMoveDecision(pokemon)
    }

    fun shouldResetwaveIndex(): Boolean {
        if (waveIndexReset) {
            waveIndexReset = false
            return true
        }
        return false
    }

    /**
     * Adds a modifier decision step to the current RL episode.
     * Each step will receive rewards when the episode terminates.
     */
    private fun addStepToCurrentEpisode(action: ModifierAction, state: SmallModifierSelectState, waveIndex: Int) {
        val currentEpisode = episodeManager.getCurrentEpisode()
        if (currentEpisode == null) {
            // No current episode - start a new one and check if it's a resumed run
            val isResumedRun = waveIndex > 1
            episodeManager.startNewEpisode(isResumedRun)

            if (isResumedRun) {
                log.warn("Started new RL episode for RESUMED run at wave {} - Training data will be marked as invalid", waveIndex)
            } else {
                log.info("Started new RL episode at wave {}", waveIndex)
            }
        }

        val reward = rewardCalculator.calculateReward(state, action)

        val step = ModifierRLStep(
            state = state,
            action = action,
            immediateReward = reward, // Small immediate rewards could be added here
            nextState = null, // Will be updated if there are multiple steps
            waveNumber = waveIndex,
            isTerminal = false,
            terminalReward = 0.0
        )

        episodeManager.addStepToCurrentEpisode(step)
        log.info(
            "Added RL step to episode: wave={}, action={}, state=HP buckets: {}",
            waveIndex, action, state.hpBuckets.contentToString()
        )
    }

    /**
     * Completes the current RL episode when a run ends.
     * This triggers terminal reward calculation and episode logging.
     */
    private fun completeCurrentEpisode(outcome: RunTerminalOutcome, waveReached: Int) {
        episodeManager.getCurrentEpisode()?.let { episode ->
            episodeManager.completeCurrentEpisode(outcome, waveReached)

            if (episode.isValidForTraining) {
                // Only log experiences for valid episodes (not resumed runs)
                val allExperiences = episode.getAllExperiences()
                allExperiences.forEach { experience ->
                    decisionLogger.logDecision(
                        experience.state,
                        experience.action,
                        experience.reward,
                        experience.nextState,
                        experience.done
                    )
                }

                log.info(
                    "Completed VALID RL episode: outcome={}, waveReached={}, steps={}, finalReward={}",
                    outcome, waveReached, episode.getStepCount(), episode.finalReward
                )

                // Save training data after each completed episode and clear buffer
                val bufferStats = decisionLogger.getBufferStats()
                val bufferSize = bufferStats["bufferSize"] as Int
                log.info(
                    "Saving training data: {} experiences from episode {}",
                    bufferSize, episodeManager.getEpisodeCount()
                )
                decisionLogger.saveAndClearBuffer()
            } else {
                log.warn(
                    "Completed INVALID RL episode (resumed run): outcome={}, waveReached={}, steps={} - EXCLUDED from training",
                    outcome, waveReached, episode.getStepCount()
                )
            }
        }
    }

    /**
     * Starts a new RL episode when a new run begins.
     * Called when the run property is created or reset.
     *
     * @param isResumedRun true if this run is resumed from a save game (waveIndex > 1)
     */
    fun startNewRLEpisode(isResumedRun: Boolean = false) {
        val newEpisode = episodeManager.startNewEpisode(isResumedRun)
        if (isResumedRun) {
            log.warn("Started new RL episode for RESUMED run: runId={} - Training data will be marked as invalid", newEpisode.runId)
        } else {
            log.info("Started new RL episode: runId={}", newEpisode.runId)
        }
    }

    /**
     * Completes the current episode with victory outcome.
     * Should be called when a run is successfully completed.
     */
    fun onRunCompleted(waveReached: Int) {
        completeCurrentEpisode(RunTerminalOutcome.VICTORY, waveReached)
    }

    /**
     * Completes the current episode with error outcome.
     * Should be called when technical errors terminate the run.
     */
    fun onRunError(waveReached: Int) {
        completeCurrentEpisode(RunTerminalOutcome.ERROR_OCCURRED, waveReached)
    }

    /**
     * Get statistics about collected RL episodes and training data.
     */
    fun getTrainingDataStats(): String {
        val loggerStats = decisionLogger.getBufferStats()
        val episodeCount = episodeManager.getEpisodeCount()
        val currentEpisode = episodeManager.getCurrentEpisode()
        val currentSteps = currentEpisode?.getStepCount() ?: 0

        return "Episodes completed: $episodeCount, Current episode steps: $currentSteps, " +
                "Buffered experiences: ${loggerStats["bufferSize"]}, " +
                "Average reward: ${loggerStats["averageReward"]}"
    }

    @Deprecated("")
    fun phaseUiIsValidated(phase: Phase, uiMode: UiMode): Boolean {
        val isValidated = longTermMemory.isUiValidated(phase)
        if (isValidated) {
            return true
        }
        if (phase is NoUiPhase) {
            longTermMemory.memorizePhase(phase.phaseName)
            return true
        }
        if (phase is UiPhase) {
            longTermMemory.memorizePhase(phase.phaseName)
            return true
        }
        return false
    }
}